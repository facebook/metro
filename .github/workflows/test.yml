name: test
on:
  workflow_call:
    inputs:
      node-version:
        type: string
        required: false
        default: 'ubuntu-latest'
      runs-on:
        type: string
        required: false
        default: '20.x'

jobs:
  test:
    name: "Tests [Node.js ${{ inputs.node-version }}, ${{ inputs.runs-on }}]"
    runs-on: ${{ inputs.runs-on }}
    steps:
      - uses: actions/checkout@v4
      - uses: ./.github/actions/yarn-install
        with:
          node-version: ${{ inputs.node-version }}
      - name: Run Jest Tests
        run: yarn jest --ci --maxWorkers 4 --reporters=default --reporters=jest-junit --rootdir='./'
