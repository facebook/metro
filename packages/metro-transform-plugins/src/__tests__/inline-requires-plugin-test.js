/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @format
 * @flow strict-local
 */

'use strict';

import type {PluginOptions, State} from '../inline-requires-plugin';
import type {PluginTesterOptions} from 'babel-plugin-tester';

const inlineRequiresPlugin = require('../inline-requires-plugin');
const validateOutputAst = require('./validateOutputAst');
const babel = require('@babel/core');
const pluginTester = require('babel-plugin-tester');
const nullthrows = require('nullthrows');

type TestCases = PluginTesterOptions<PluginOptions, State>['tests'];

const TEST_CASES: TestCases = {
  'inlines single usage': {
    code: ['var foo = require("foo");', 'foo.bar()'].join('\n'),
    snapshot: true,
  },

  'inlines multiple usages': {
    code: ['var foo = require("foo");', 'foo.bar()', 'foo.baz()'].join('\n'),
    snapshot: true,
  },

  'inlines any number of variable declarations': {
    code: [
      'var foo = require("foo"), bar = require("bar"), baz = 4;',
      'foo.method()',
    ].join('\n'),
    snapshot: true,
  },

  'ignores requires that are not assigned': {
    code: ['require("foo");'].join('\n'),
    snapshot: false,
  },

  'delete unused requires': {
    code: ['var foo = require("foo");'].join('\n'),
    snapshot: true,
  },

  'ignores requires that are re-assigned': {
    code: ['var foo = require("foo");', 'foo = "bar";'].join('\n'),
    snapshot: false,
  },

  'ensures that the inlined require still points to the global require function':
    {
      code: `
          const foo = require('foo');

          function test() {
            function require(condition) {
              if (!condition) {
                throw new Error('Condition is falsy');
              }
            }

            require(foo.isOnline());
          }
        `,
      snapshot: true,
    },

  'ensures that the inlined require still points to the global require function with inlineableCalls options':
    {
      code: `
          const foo = customStuff('foo');

          function test() {
            function customStuff(condition) {
              if (!condition) {
                throw new Error('Condition is falsy');
              }
            }

            customStuff(foo.isOnline());
          }
        `,
      snapshot: true,
    },

  'ensures that the inlined require still points to the global require function even if local require is not called':
    {
      code: `
          const foo = require('foo');

          function test() {
            function require(condition) {
              if (!condition) {
                throw new Error('Condition is falsy');
              }
            }

            foo.isOnline();
          }
        `,
      snapshot: true,
    },

  'does not transform require calls if require is redeclared in the same declaration scope':
    {
      code: `
          function require(condition) {
            if (!condition) {
              throw new Error('Condition is falsy');
            }
          }
          const foo = require('foo');
          console.log(foo.test);
        `,
      snapshot: false,
    },

  'does not transform require calls if require is redeclared in the global scope':
    {
      code: `
          function require(condition) {
            if (!condition) {
              throw new Error('Condition is falsy');
            }
          }
          function test() {
            const foo = require('foo');
            console.log(foo.test);
          }
        `,
      snapshot: false,
    },

  'does not transform require calls that are already inline': {
    code: `
        function test() {
          function require(condition) {
            if (!condition) {
              throw new Error('The condition is false');
            }
          }
          require('test');
        }
      `,
    snapshot: false,
  },

  'inlines requires that are referenced before the require statement': {
    code: [
      'function foo() {',
      '  bar();',
      '}',
      'var bar = require("baz");',
      'foo();',
      'bar();',
    ].join('\n'),
    snapshot: true,
  },

  'inlines require properties': {
    code: [
      'var tmp = require("./a");',
      'var a = tmp.a',
      'var D = {',
      '  b: function(c) { c ? a(c.toString()) : a("No c!"); },',
      '};',
    ].join('\n'),
    snapshot: true,
  },

  'ignores require properties (as identifiers) that are re-assigned': {
    code: [
      'var X = require("X");',
      'var origA = X.a',
      'X.a = function() {',
      '  origA();',
      '};',
    ].join('\n'),
    snapshot: true,
  },

  'ignores require properties (as strings) that are re-assigned': {
    code: [
      'var X = require("X");',
      'var origA = X["a"]',
      'X["a"] = function() {',
      '  origA();',
      '};',
    ].join('\n'),
    snapshot: true,
  },

  'inlines functions provided via `inlineableCalls`': {
    code: [
      'const inlinedCustom = customStuff("foo");',
      'const inlinedRequire = require("bar");',
      '',
      'inlinedCustom();',
      'inlinedRequire();',
    ].join('\n'),
    snapshot: true,
  },

  'ignores requires in `ignoredRequires`': {
    code: ['const CommonFoo = require("CommonFoo");', 'CommonFoo();'].join(
      '\n',
    ),
    snapshot: false,
  },

  'ignores destructured properties of requires in `ignoredRequires`': {
    code: [
      'const tmp = require("CommonFoo");',
      'const a = require("CommonFoo").a;',
      'a();',
    ].join('\n'),
    snapshot: false,
  },

  'inlines require.resolve calls': {
    code: ['const a = require(require.resolve("Foo")).bar;', '', 'a();'].join(
      '\n',
    ),
    snapshot: true,
  },

  'inlines with multiple arguments': {
    code: ['const a = require("Foo", "Bar", 47);', '', 'a();'].join('\n'),
    snapshot: true,
  },
};

describe.each([true, false])('memoizeCalls=%s:', memoizeCalls => {
  pluginTester<PluginOptions, State>({
    babelOptions: {
      babelrc: false,
      configFile: false,
    },
    plugin: inlineRequiresPlugin,
    pluginOptions: {
      ignoredRequires: ['CommonFoo'],
      inlineableCalls: ['customStuff'],
      memoizeCalls,
    },
    tests: TEST_CASES,
  });
});

describe('inline-requires', () => {
  const transform = (
    source: $ReadOnlyArray<string>,
    options?: $ReadOnly<{...}>,
  ) =>
    babel.transformSync(source.join('\n'), {
      ast: true,
      compact: true,
      plugins: [
        // $FlowFixMe[untyped-import] @babel/plugin-transform-modules-commonjs
        [require('@babel/plugin-transform-modules-commonjs'), {strict: false}],
        [inlineRequiresPlugin, options],
      ],
    });

  const compare = (
    input: $ReadOnlyArray<string>,
    output: $ReadOnlyArray<string>,
    options?: $ReadOnly<{...}>,
  ) => {
    expect(transform(input, options).code).toBe(
      transform(output, options).code,
    );
  };

  it('should be compatible with other transforms like transform-modules-commonjs', function () {
    compare(
      ['import Imported from "foo";', 'console.log(Imported);'],
      [
        'var _foo = _interopRequireDefault(require("foo"));',
        'function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }',
        'console.log(_foo.default);',
      ],
    );
  });

  it('should be compatible with `transform-modules-commonjs` when using named imports', function () {
    compare(
      [
        'import {a} from "./a";',
        'var D = {',
        '  b: function(c) { c ? a(c.toString()) : a("No c!"); },',
        '};',
      ],
      [
        'var D = {',
        '  b: function (c) {',
        '    c ? (0, require("./a").a)(c.toString()) : (0, require("./a").a)("No c!");',
        '  }',
        '};',
      ],
    );
  });

  it('should remove loc information from nodes', function () {
    const ast = transform(['var x = require("x"); x']).ast;
    expect(ast).not.toBeNull();
    const expression = nullthrows(ast).program.body[0].expression;

    function expectNodeWithNoLocation(maybeNode: ?BabelNode) {
      expect(maybeNode).not.toBeNull();
      const node = nullthrows(maybeNode);
      expect(node.start).toBeUndefined();
      expect(node.end).toBeUndefined();
      expect(node.loc).toBeUndefined();
    }

    expectNodeWithNoLocation(expression);
    expectNodeWithNoLocation(nullthrows(expression?.arguments)[0]);
  });

  it('should not emit duplicate nodes', function () {
    const ast = transform([
      'var foo = require("foo");',
      'foo.bar()',
      'foo.baz()',
    ]).ast;
    validateOutputAst(nullthrows(ast));
  });
});
