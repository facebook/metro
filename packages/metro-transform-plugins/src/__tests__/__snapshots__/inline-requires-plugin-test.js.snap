// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`memoizeCalls=false: inline-requires delete unused requires: delete unused requires 1`] = `
"
var foo = require(\\"foo\\");

      ↓ ↓ ↓ ↓ ↓ ↓


"
`;

exports[`memoizeCalls=false: inline-requires ensures that the inlined require still points to the global require function even if local require is not called: ensures that the inlined require still points to the global require function even if local require is not called 1`] = `
"
const foo = require('foo');

function test() {
  function require(condition) {
    if (!condition) {
      throw new Error('Condition is falsy');
    }
  }

  foo.isOnline();
}

      ↓ ↓ ↓ ↓ ↓ ↓

function test() {
  function _require(condition) {
    if (!condition) {
      throw new Error('Condition is falsy');
    }
  }
  require('foo').isOnline();
}
"
`;

exports[`memoizeCalls=false: inline-requires ensures that the inlined require still points to the global require function with inlineableCalls options: ensures that the inlined require still points to the global require function with inlineableCalls options 1`] = `
"
const foo = customStuff('foo');

function test() {
  function customStuff(condition) {
    if (!condition) {
      throw new Error('Condition is falsy');
    }
  }

  customStuff(foo.isOnline());
}

      ↓ ↓ ↓ ↓ ↓ ↓

function test() {
  function _customStuff(condition) {
    if (!condition) {
      throw new Error('Condition is falsy');
    }
  }
  _customStuff(customStuff('foo').isOnline());
}
"
`;

exports[`memoizeCalls=false: inline-requires ensures that the inlined require still points to the global require function: ensures that the inlined require still points to the global require function 1`] = `
"
const foo = require('foo');

function test() {
  function require(condition) {
    if (!condition) {
      throw new Error('Condition is falsy');
    }
  }

  require(foo.isOnline());
}

      ↓ ↓ ↓ ↓ ↓ ↓

function test() {
  function _require(condition) {
    if (!condition) {
      throw new Error('Condition is falsy');
    }
  }
  _require(require('foo').isOnline());
}
"
`;

exports[`memoizeCalls=false: inline-requires ignores require properties (as identifiers) that are re-assigned: ignores require properties (as identifiers) that are re-assigned 1`] = `
"
var X = require(\\"X\\");
var origA = X.a
X.a = function() {
  origA();
};

      ↓ ↓ ↓ ↓ ↓ ↓

var origA = require(\\"X\\").a;
require(\\"X\\").a = function () {
  origA();
};
"
`;

exports[`memoizeCalls=false: inline-requires ignores require properties (as strings) that are re-assigned: ignores require properties (as strings) that are re-assigned 1`] = `
"
var X = require(\\"X\\");
var origA = X[\\"a\\"]
X[\\"a\\"] = function() {
  origA();
};

      ↓ ↓ ↓ ↓ ↓ ↓

var origA = require(\\"X\\")[\\"a\\"];
require(\\"X\\")[\\"a\\"] = function () {
  origA();
};
"
`;

exports[`memoizeCalls=false: inline-requires inlines any number of variable declarations: inlines any number of variable declarations 1`] = `
"
var foo = require(\\"foo\\"), bar = require(\\"bar\\"), baz = 4;
foo.method()

      ↓ ↓ ↓ ↓ ↓ ↓

var baz = 4;
require(\\"foo\\").method();
"
`;

exports[`memoizeCalls=false: inline-requires inlines functions provided via \`inlineableCalls\`: inlines functions provided via \`inlineableCalls\` 1`] = `
"
const inlinedCustom = customStuff(\\"foo\\");
const inlinedRequire = require(\\"bar\\");

inlinedCustom();
inlinedRequire();

      ↓ ↓ ↓ ↓ ↓ ↓

customStuff(\\"foo\\")();
require(\\"bar\\")();
"
`;

exports[`memoizeCalls=false: inline-requires inlines multiple usages: inlines multiple usages 1`] = `
"
var foo = require(\\"foo\\");
foo.bar()
foo.baz()

      ↓ ↓ ↓ ↓ ↓ ↓

require(\\"foo\\").bar();
require(\\"foo\\").baz();
"
`;

exports[`memoizeCalls=false: inline-requires inlines require properties: inlines require properties 1`] = `
"
var tmp = require(\\"./a\\");
var a = tmp.a
var D = {
  b: function(c) { c ? a(c.toString()) : a(\\"No c!\\"); },
};

      ↓ ↓ ↓ ↓ ↓ ↓

var D = {
  b: function (c) {
    c ? require(\\"./a\\").a(c.toString()) : require(\\"./a\\").a(\\"No c!\\");
  }
};
"
`;

exports[`memoizeCalls=false: inline-requires inlines require.resolve calls: inlines require.resolve calls 1`] = `
"
const a = require(require.resolve(\\"Foo\\")).bar;

a();

      ↓ ↓ ↓ ↓ ↓ ↓

require(require.resolve(\\"Foo\\")).bar();
"
`;

exports[`memoizeCalls=false: inline-requires inlines requires that are referenced before the require statement: inlines requires that are referenced before the require statement 1`] = `
"
function foo() {
  bar();
}
var bar = require(\\"baz\\");
foo();
bar();

      ↓ ↓ ↓ ↓ ↓ ↓

function foo() {
  require(\\"baz\\")();
}
foo();
require(\\"baz\\")();
"
`;

exports[`memoizeCalls=false: inline-requires inlines single usage: inlines single usage 1`] = `
"
var foo = require(\\"foo\\");
foo.bar()

      ↓ ↓ ↓ ↓ ↓ ↓

require(\\"foo\\").bar();
"
`;

exports[`memoizeCalls=false: inline-requires inlines with multiple arguments: inlines with multiple arguments 1`] = `
"
const a = require(\\"Foo\\", \\"Bar\\", 47);

a();

      ↓ ↓ ↓ ↓ ↓ ↓

require(\\"Foo\\", \\"Bar\\", 47)();
"
`;

exports[`memoizeCalls=true: inline-requires delete unused requires: delete unused requires 1`] = `
"
var foo = require(\\"foo\\");

      ↓ ↓ ↓ ↓ ↓ ↓


"
`;

exports[`memoizeCalls=true: inline-requires ensures that the inlined require still points to the global require function even if local require is not called: ensures that the inlined require still points to the global require function even if local require is not called 1`] = `
"
const foo = require('foo');

function test() {
  function require(condition) {
    if (!condition) {
      throw new Error('Condition is falsy');
    }
  }

  foo.isOnline();
}

      ↓ ↓ ↓ ↓ ↓ ↓

var foo;
function test() {
  function _require(condition) {
    if (!condition) {
      throw new Error('Condition is falsy');
    }
  }
  (foo || (foo = require('foo'))).isOnline();
}
"
`;

exports[`memoizeCalls=true: inline-requires ensures that the inlined require still points to the global require function with inlineableCalls options: ensures that the inlined require still points to the global require function with inlineableCalls options 1`] = `
"
const foo = customStuff('foo');

function test() {
  function customStuff(condition) {
    if (!condition) {
      throw new Error('Condition is falsy');
    }
  }

  customStuff(foo.isOnline());
}

      ↓ ↓ ↓ ↓ ↓ ↓

var foo;
function test() {
  function _customStuff(condition) {
    if (!condition) {
      throw new Error('Condition is falsy');
    }
  }
  _customStuff((foo || (foo = customStuff('foo'))).isOnline());
}
"
`;

exports[`memoizeCalls=true: inline-requires ensures that the inlined require still points to the global require function: ensures that the inlined require still points to the global require function 1`] = `
"
const foo = require('foo');

function test() {
  function require(condition) {
    if (!condition) {
      throw new Error('Condition is falsy');
    }
  }

  require(foo.isOnline());
}

      ↓ ↓ ↓ ↓ ↓ ↓

var foo;
function test() {
  function _require(condition) {
    if (!condition) {
      throw new Error('Condition is falsy');
    }
  }
  _require((foo || (foo = require('foo'))).isOnline());
}
"
`;

exports[`memoizeCalls=true: inline-requires ignores require properties (as identifiers) that are re-assigned: ignores require properties (as identifiers) that are re-assigned 1`] = `
"
var X = require(\\"X\\");
var origA = X.a
X.a = function() {
  origA();
};

      ↓ ↓ ↓ ↓ ↓ ↓

var X;
var origA = (X || (X = require(\\"X\\"))).a;
X.a = function () {
  origA();
};
"
`;

exports[`memoizeCalls=true: inline-requires ignores require properties (as strings) that are re-assigned: ignores require properties (as strings) that are re-assigned 1`] = `
"
var X = require(\\"X\\");
var origA = X[\\"a\\"]
X[\\"a\\"] = function() {
  origA();
};

      ↓ ↓ ↓ ↓ ↓ ↓

var X;
var origA = (X || (X = require(\\"X\\")))[\\"a\\"];
X[\\"a\\"] = function () {
  origA();
};
"
`;

exports[`memoizeCalls=true: inline-requires inlines any number of variable declarations: inlines any number of variable declarations 1`] = `
"
var foo = require(\\"foo\\"), bar = require(\\"bar\\"), baz = 4;
foo.method()

      ↓ ↓ ↓ ↓ ↓ ↓

var foo;
var baz = 4;
(foo || (foo = require(\\"foo\\"))).method();
"
`;

exports[`memoizeCalls=true: inline-requires inlines functions provided via \`inlineableCalls\`: inlines functions provided via \`inlineableCalls\` 1`] = `
"
const inlinedCustom = customStuff(\\"foo\\");
const inlinedRequire = require(\\"bar\\");

inlinedCustom();
inlinedRequire();

      ↓ ↓ ↓ ↓ ↓ ↓

var inlinedRequire;
var inlinedCustom;
(inlinedCustom || (inlinedCustom = customStuff(\\"foo\\")))();
(inlinedRequire || (inlinedRequire = require(\\"bar\\")))();
"
`;

exports[`memoizeCalls=true: inline-requires inlines multiple usages: inlines multiple usages 1`] = `
"
var foo = require(\\"foo\\");
foo.bar()
foo.baz()

      ↓ ↓ ↓ ↓ ↓ ↓

var foo;
(foo || (foo = require(\\"foo\\"))).bar();
foo.baz();
"
`;

exports[`memoizeCalls=true: inline-requires inlines require properties: inlines require properties 1`] = `
"
var tmp = require(\\"./a\\");
var a = tmp.a
var D = {
  b: function(c) { c ? a(c.toString()) : a(\\"No c!\\"); },
};

      ↓ ↓ ↓ ↓ ↓ ↓

var tmp;
var a = (tmp || (tmp = require(\\"./a\\"))).a;
var D = {
  b: function (c) {
    c ? a(c.toString()) : a(\\"No c!\\");
  }
};
"
`;

exports[`memoizeCalls=true: inline-requires inlines require.resolve calls: inlines require.resolve calls 1`] = `
"
const a = require(require.resolve(\\"Foo\\")).bar;

a();

      ↓ ↓ ↓ ↓ ↓ ↓

var a;
(a || (a = require(require.resolve(\\"Foo\\")).bar))();
"
`;

exports[`memoizeCalls=true: inline-requires inlines requires that are referenced before the require statement: inlines requires that are referenced before the require statement 1`] = `
"
function foo() {
  bar();
}
var bar = require(\\"baz\\");
foo();
bar();

      ↓ ↓ ↓ ↓ ↓ ↓

var bar;
function foo() {
  (bar || (bar = require(\\"baz\\")))();
}
foo();
(bar || (bar = require(\\"baz\\")))();
"
`;

exports[`memoizeCalls=true: inline-requires inlines single usage: inlines single usage 1`] = `
"
var foo = require(\\"foo\\");
foo.bar()

      ↓ ↓ ↓ ↓ ↓ ↓

var foo;
(foo || (foo = require(\\"foo\\"))).bar();
"
`;

exports[`memoizeCalls=true: inline-requires inlines with multiple arguments: inlines with multiple arguments 1`] = `
"
const a = require(\\"Foo\\", \\"Bar\\", 47);

a();

      ↓ ↓ ↓ ↓ ↓ ↓

var a;
(a || (a = require(\\"Foo\\", \\"Bar\\", 47)))();
"
`;
